crates/bevy_core/src/time/timer.rs

```diff
@@ -364,6 +367,334 @@ impl Timer {
     }
 }
 
+/// Tracks elapsed increments. Enters the finished state once `duration` is reached.
+///
+/// This type is useful for accurately counting frames, ticks or occurences between gameplay events.
+/// Use ['Timer'] if you care about wall-clock time.
+///
+/// Non-repeating timers will stop tracking and stay in the finished state until reset.
+/// Repeating timers will only be in the finished state on each tick `duration` is reached or
+/// exceeded, and can still be reset at any given point.
+///
+/// Paused timers will not have elapsed time increased.
+#[derive(Clone, Debug, Default, Reflect)]
+#[reflect(Component)]
+pub struct DiscreteTimer {
+    stopwatch: DiscreteStopwatch,
+    duration: u64,
+    repeating: bool,
+    finished: bool,
+    times_finished: u32,
+}
+
+impl DiscreteTimer {
+    /// Creates a new timer with a given duration.
+    pub fn new(duration: u64, repeating: bool) -> Self {
+        Self {
+            duration,
+            repeating,
+            ..Default::default()
+        }
+    }
+
+    /// Returns `true` if the timer has reached its duration.
+    ///
+    /// # Examples
+    /// ```
+    /// # use bevy_core::*;
+    /// let mut timer = DiscreteTimer::new(3, false);
+    /// timer.tick(3);
+    /// assert!(timer.finished());
+    /// timer.tick(1);
+    /// assert!(timer.finished());
+    /// ```
+    #[inline]
+    pub fn finished(&self) -> bool {
+        self.finished
+    }
+
+    /// Returns `true` only on the tick the timer reached its duration.
+    ///
+    /// # Examples
+    /// ```
+    /// # use bevy_core::*;
+    /// let mut timer = DiscreteTimer::new(2, false);
+    /// timer.tick(2);
+    /// assert!(timer.just_finished());
+    /// timer.tick(1);
+    /// assert!(!timer.just_finished());
+    /// ```
+    #[inline]
+    pub fn just_finished(&self) -> bool {
+        self.times_finished > 0
+    }
+
+    /// Returns the time elapsed on the timer. Guaranteed to be between 0 and `duration`.
+    /// Will only equal `duration` when the timer is finished and non repeating.
+    ///
+    /// See also [`DiscreteStopwatch::elapsed`](DiscreteStopwatch::elapsed).
+    ///
+    /// # Examples
+    /// ```
+    /// # use bevy_core::*;
+    /// let mut timer = DiscreteTimer::new(2, false);;
+    /// timer.tick(1);
+    /// assert_eq!(timer.elapsed(), 1);
+    /// ```
+    #[inline]
+    pub fn elapsed(&self) -> u64 {
+        self.stopwatch.elapsed()
+    }
+
+    /// Sets the elapsed time of the timer without any other considerations.
+    ///
+    /// See also [`Stopwatch::set`](Stopwatch::set).
+    ///
+    /// #
+    /// ```
+    /// # use bevy_core::*;
+    /// use std::time::Duration;
+    /// let mut timer = Timer::from_seconds(1.0, false);
+    /// timer.set_elapsed(Duration::from_secs(2));
+    /// assert_eq!(timer.elapsed(), Duration::from_secs(2));
+    /// // the timer is not finished even if the elapsed time is greater than the duration.
+    /// assert!(!timer.finished());
+    /// ```
```

Comment:
should be
```rust
    /// ```
    /// # use bevy_core::*;
    /// let mut timer = DiscreteTimer::new(1, false);
    /// timer.set_elapsed(2);
    /// assert_eq!(timer.elapsed(), 2);
    /// // The timer is not finished, even if the elapsed time is greater than the duration.
    /// assert!(!timer.finished());
    /// ```
```